### DISCUSSION TITLE
Ethernet Preemption - Showcase

### DISCUSSION THREAD

**Discussion by adamgeorge309** [inet-framework/inet#676] [Showcases]
Discuss the "Ethernet Preemption" showcase here. Ideas and recommendations are welcome.

**Comment by levy** [comment]
It might be true that this combination doesn't work in TsnDevice/TsnSwitch
out of the box. Please check the example under
showcases/tsn/framepreemption to see where the queue is in the preempting
MAC layer. The queue needs to be placed under the sub MAC layers in order
to operate correctly. Other than that, I think it should be fine.

Best regards,
levy

**Reply by pablomlu** [reply]
I also encountered the same problem when activating time-aware shaping and frame preemption in TsnSwitch. The packets do not seem to go through the network. Is there any solution for this? Thanks in advance.

**Reply by levy** [reply]
Checkout the latest INET master branch and see the showcases/combiningfeatures/fpandtas example.

**Reply by pablomlu** [reply]
Thanks a lot, very interesting example. It seems that the example works fine with two traffic classes. I wonder if it would be possible in the example to have an array of preemptableMacLayer.gate to have multiples gates or queues.  Do you know if it this scenario can be configured? Thanks again.

**Reply by gimmyy** [reply]
Thanks for your great job! Would it be possible to provide a showcase with mixed features of FramePreemption, TAS, and CBS. It would conform to the TSN preemption mode even more. Thank you so much.

**Reply by anoikob** [reply]
> Checkout the latest INET master branch and see the showcases/combiningfeatures/fpandtas example.

Ι copy pasted the ini code in a new ini file, under combiningfeatures/fpandtas (also a new folder) and got this error. I'm using omnet 6.0.3 and inet 4.5. Any ideas?
![image](https://github.com/user-attachments/assets/8433476a-1d83-4556-9cc9-5b0be95d5f80)

**Comment by levy** [comment]
Once I had a working example which contained multiple preemptable MAC layers. A lower priority MAC layer could be preempted by a higher priority one, even in a nested way where A is preempted by B, then B is preempted by C, then C finishes, then B finishes, and finally A finishes. IIRC, the complication is at the reassembly at the receiver, because it should be prepared for this nested preemption. There could be multiple frames preempted at different priorities and the received fragments should not be mixed up. Anyway, it's not possible to simply configure such a scenario right now, but with a bit of programming it could work.

**Reply by pablomlu** [reply]
Ok, I have modified some lines in the initialization file and I got the desired configuration working. Preemption in a nested way is more complex than my target scenario. Thank you!

**Reply by jafficy** [reply]
Hello, I also encountered the same problem, the traffic does not seem to enter the MAC sub-layer, where are you modified, to complete the mixed use of TAS and frame preemption。Thank you！

**Reply by pablomlu** [reply]
Hi, just follow the showcases/combiningfeatures/fpandtas example available in the INET master branch and the mixed use of TAS and frame preemption should work.

**Reply by pablomlu** [reply]
The example can be found at github.com/inet-framework/inet/tree/master/showcases/tsn/combiningfeatures

**Reply by jafficy** [reply]
Thank you very much and have a great day！

**Reply by jafficy** [reply]
Excuse me again，When I run the fpandtas example .The following problems have arisen： Submodule queue: No module type named 'GatingQueue' found that implements module interface inet.queueing.contract.IPacketQueue (not in the loaded NED files?), at /home/bx/Desktop/omnetpp-6.0.1/samples/inet4.5/src/inet/linklayer/ethernet/modular/EthernetStreamingMacLayer.ned:37 -- in module (omnetpp::cModule) TsnLinearNetwork.client.eth[0].macLayer.expressMacLayer (id=116), during network setup

**Reply by levy** [reply]
This example is working in INET master only, it's not included in older versions.

**Comment by jafficy** [comment]


**Reply by jafficy** [reply]
Thank you very much for your reply！I'm running on the latest version with inet4.5.2,But the error is displayed,Which version can I use?

**Reply by levy** [reply]
I have already said: use the INET master branch from the GitHub repository.
